storage-proofs/src/compound_proof.rs

```diff
@@ -122,11 +122,13 @@ where
         multi_proof: &MultiProof<'b, E>,
         requirements: &S::Requirements,
     ) -> Result<bool> {
+        ensure!(
+            multi_proof.circuit_proofs.len() == Self::partition_count(public_params),
+            "Inconsistent inputs"
+        );
+
         let vanilla_public_params = &public_params.vanilla_params;
-        let pvk = groth16::prepare_verifying_key(&multi_proof.verifying_key);
-        if multi_proof.circuit_proofs.len() != Self::partition_count(public_params) {
-            return Ok(false);
-        }
+        let pvk = groth16::prepare_batch_verifying_key(&multi_proof.verifying_key);
```

Comment:
I guess we could cache this and avoid incurring the cost on every batch verification, too.

Reply:
We could, but that would require leaking this outside of the proofs code